src/python/pants/backend/docker/docker_build_context.py

```diff
@@ -30,19 +34,34 @@ class DockerBuildContext:
 @dataclass(frozen=True)
 class DockerBuildContextRequest:
     address: Address
-    context_root: str
+    build_upstream_images: bool = False
 
 
 @rule
 async def create_docker_build_context(request: DockerBuildContextRequest) -> DockerBuildContext:
+    # Get all targets to include in context.
     transitive_targets = await Get(TransitiveTargets, TransitiveTargetsRequest([request.address]))
 
-    # Get all sources, going into the build context.
+    # Get all dependencies from those root targets.
+    all_dependencies = await MultiGet(
+        Get(Targets, DependenciesRequest(target.get(Dependencies), include_special_cased_deps=True))
+        for target in transitive_targets.roots
+    )
+
+    # Get Dockerfiles from all roots.
+    dockerfiles_request = Get(
+        SourceFiles,
+        SourceFilesRequest(
+            sources_fields=[t.get(Sources) for t in transitive_targets.roots],
+            for_sources_types=(DockerImageSources,),
+        ),
+    )
+    # Get all sources from all dependencies (i.e. files and resources).
     sources_request = Get(
         SourceFiles,
         SourceFilesRequest(
-            sources_fields=[t.get(Sources) for t in transitive_targets.closure],
-            for_sources_types=(DockerImageSources, FilesSources, ResourcesSources),
+            sources_fields=[t.get(Sources) for t in chain(*all_dependencies)],
+            for_sources_types=(FilesSources, ResourcesSources),
```

Comment:
I think we might as well go ahead and stop handling `ResourcesSources` here. As we discussed, it doesn't make much sense to do so.

Reply:
Hmm, and we will still pull in the `files()` targets that we depend on via other `docker_image` targets, but as previously discussed this is a general problem that we don't need to solve here right now.

Reply:
Check on dropping resources here.

No, I have tests covering that they don't pull in `files` from other `docker_image`s.

Reply:
What about archives?

Reply:
I think those should just work - they are packaged resources. Are you seeing otherwise?

I'm surprised that files from other images (that this one depends on) aren't being pulled in. What prevents it?

Reply:
I haven't tried anything with archives yet. Great if it just works. :) 

`all_dependencies` are not really *all* dependencies for all transient targets, just all dependencies for the roots, i.e. our docker_file.

Reply:
Ah, right. `packaged` resources, now I get it. Awesome. Yeah, that should work.

Reply:
I rephrased the comments and renamed a few variables to make things hopefully clearer.